### DISCUSSION TITLE
posts/environment-variables-in-react-native-the-right-way/

### DISCUSSION THREAD

**Discussion by giscus** [yjose/elazizi.com#99] [Blog posts]
# posts/environment-variables-in-react-native-the-right-way/

In this article, I will show you how to use environment variables in React Native the correct way.

https://elazizi.com/posts/environment-variables-in-react-native-the-right-way/

**Comment by lucksp** [comment]
This is a nice write up. How do you handle ENV secrets when using EAS to build?  I don’t think you can actually use a local env file within that build environment

**Reply by yjose** [reply]
Thank you @lucksp 🙏, glad you find it helpful, 
Regarding your question, you can use whatever method you choose based on the [Official docs](https://docs.expo.dev/build-reference/variables/). But, I think the best option that I have found very handy is pushing the `.env.production` with my Git repository, but without all critical keys.
So, I push my prod env file with keys such as bundle ID and app name, and add other keys using the command line.
```
eas secret:create --scope project --name SECRET_NAME --value secretvalue --type string
```
You can also add your entire `.env.production` file as a secret to EAS things, but it's not handy as you need to update it whenever you add a new environment variable.

**Reply by lucksp** [reply]
Ok. So just keep the secrets out and handle those separately. Thanks!!

**Reply by ovidb** [reply]
I have a question, if you use eas secrets, how do you validate them? Furthermore, how do you import them?

**Comment by AlexLvovsky** [comment]
Great article!
One question: is it "bad practice" to use Firebase RemoteConfig for example?

**Reply by yjose** [reply]
Thank you @AlexLvovsky, I'm not sure how your question relates to the article, but anyway, from my understanding, Firebase Remote Config is a tool for A/B testing on mobile apps, so it mainly works at runtime. The approach I'm sharing in this article is perfect for most use cases when you need to support multiple variants of the app (dev/staging/prod), but if you want feature flag support, you need to use tools such as Firebase Remote Config.

Anyway, let me know if this answers your question. If it's not the case, please elaborate more on your scenario.

**Comment by ovidb** [comment]
I have a question, if you use eas secrets, how do you validate them? Furthermore, how do you import them?

**Reply by yjose** [reply]
Normally, secrets are injected into `process.env` automatically, just like environment variables added to the `eas.json` file. So, all your secrets will be available during the build validation phase while running the build.

but to be honest, I am a little bit concerned about the last updates with the Expo 49 SDK, which introduced loading environment files automatically, and I still haven't tested it enough to have a complete understanding.

if you try it would love to hear your feedback

**Comment by rimancete** [comment]
Nice post bro!

Thanks for sharing

**Comment by yjose** [comment]
🚨 If you're experiencing issues with Ecpo SDK 49 due to the new automatic env file loading feature, I recommend temporarily disabling it by using `EXPO_NO_DOTENV=1` before running your scripts. This will disable the new feature.

I am still experimenting with how we can make this approach work without any issues.

**Comment by kmml0** [comment]
How to resolve this error? Its failing to use "import" in a "js" file
SyntaxError: Error reading Expo config at /Users/.../app.config.ts:
Cannot use import statement outside a module
/Users/.../env.js:8
import Constants from 'expo-constants'
^^^^^^

**Reply by yjose** [reply]
Not sure exactly what the issue is. If you can share the code, it would be easy to give you more details. By the way, comments support Markdown.
But from what you wrote, it looks like you are trying to use an import statement inside `env.js`. This will throw an error. However, you can use `const Constants = require('expo-constants')` instead.

**Reply by kmml0** [reply]
Yeah, Im using solito, and so far just copied the `env.js` file from your repo (`core/env.js`) which uses an import statement.

**Reply by yjose** [reply]
Okay, if you are talking about `core/env.js`, you can actually use the import statement.
I believe you are importing `core/env.js` into `app.config.ts`, which is not what we explained in the article.  you should import `./env.js` instead

**Reply by kmml0** [reply]
I was referring to this [file](https://github.com/obytes/react-native-template-obytes/blob/master/src/core/env.js) and this import statement: `import Constants from 'expo-constants'` which is failing with `Cannot use import statement outside a module`

**Reply by kmml0** [reply]
Nvm, fixed. Thanks for the great tutorial!

**Comment by vbuongiovanni** [comment]
This solution works beautifully for those using expo prebuild... Just be sure to clean and rebuild your project after making these changes, and any other changes associated with your .env's

Thank you, Youssouf! You are a gentleman and a scholar.

**Reply by yjose** [reply]
Thank you @vbuongiovanni, Glad you found it useful.

**Comment by Obiora-Imah** [comment]
How does this whole setup work with react native testing library test and jest?

**Comment by juliusoh** [comment]
Not sure why im getting

Error: Error reading Expo config at /app.config.ts:
Cannot find module './env'

**Reply by teampazi** [reply]
> Not sure why im getting
> 
> Error: Error reading Expo config at /app.config.ts: Cannot find module './env'

Try prefixing your var with  `EXPO_`

**Comment by mrevanzak** [comment]
i follow all off them but got 
`❌ Metro encountered an error:
Unable to resolve module path ` because of importing 'path' module. even if i install it as a dev deps, it will shows another error that comes from dotenv

**Reply by yjose** [reply]
Make sure you are not importing from ./env  innside src files instead use @env

**Reply by mrevanzak** [reply]
> Make sure you are not importing from ./env innside src files instead use @env

yes i double checked that. its weird

**Reply by mrevanzak** [reply]
<img width="410" alt="image" src="https://github.com/user-attachments/assets/2207d267-3a6f-471c-97da-642104ee176f">
see

**Reply by yjose** [reply]
"Unable to resolve module path" means you are trying to import the `path` Node.js module into your app, which will not work because the runtime is different. This issue arises because you are importing the `./env.js` file into a file located in the `./src` folder, which Metro is attempting to bundle. 

To resolve this, ensure you import your environment variables from `@env`, which should be an [alias](https://github.com/obytes/react-native-template-obytes/blob/a46693e6ddad22bd42e8c03fec623120d9d376d4/babel.config.js#L15) for `./src/core/env.js`, instead of `./env.js`.

let's me know if the issue still persist after those checks

**Reply by mrevanzak** [reply]
yes i already look into it. see my screenshot above. that's all env that i imported and all cames from "@env". it feels weird for me too. why it is so hard to manage env

**Reply by yjose** [reply]
Did you check your Babel config, and the aliases are already there?

**Reply by mrevanzak** [reply]
yes. i also tried with removing the alias path and make sure imported the one that used expo constants

**Comment by yalipanerga** [comment]
have you test this with eas build local ?

**Reply by yjose** [reply]
yeah, before running on local make sure to export the APP_ENV with the correct value before running the build

**Comment by DeveloperInside** [comment]
It works in the debug build but not in the release build on Xcode. I’m looking for a solution but haven’t found one yet.

**Reply by yjose** [reply]
For xcode you should export your environments variable before running the build tge same way we do in [EAS](https://github.com/obytes/react-native-template-obytes/blob/master/eas.json).

**Comment by azael20** [comment]
Hello, I followed all the steps and my application runs without errors. The issue is that with Expo-Constants, it returns the values of the variables, but when I try to use @env to have TypeScript support, it doesn't return any value (for example, Env.APP_ENV). I hope you can help me solve this issue. Greetings
